Import / Export & Transaction Log features

mankyKitty opened the pull request:
This branch contains both the import/export feature and the transaction log feature.

The import/export has been extended to include the transaction logs.

Supersedes #546 & #541 

- [x] fix merge conflicts
- [x] test all builds
  - [X] deb
  - [X] ova
  - [x] mac
  - [x] ghcjs

benkolera committed:
Something that compiles that Sean can point to :)

benkolera committed:
Inversion of back workflow. Call file dialog from import page

benkolera committed:
Adds file a file type param to the open file dialog

Not wired into mac yet

benkolera committed:
Import form functional. Needs styling and impl of doImport

benkolera committed:
import Is mostly working

Something is overwriting what we restore. Needs error checking and styling.
And mac / linux deliver file support

benkolera committed:
Remove Debug.Trace

benkolera committed:
linux import export working nicely

benkolera committed:
Mac dialogs mostly working. Not quite, but close enough to push

benkolera committed:
Merge branch 'develop' into blk@develop/import_export

benkolera committed:
Auto popup the file dialog on import

benkolera committed:
Show import dialog on import page load

benkolera committed:
Import styling and error handling done

benkolera committed:
Merge branch 'blk@develop/import_export' of github.com:kadena-io/chainweaver into blk@develop/import_export

benkolera committed:
Only show filename for import file

benkolera committed:
Mac working, yay

benkolera committed:
Fixed linux for save dialog stuff

Commit:
Merged new Transaction Logs to Import/Export feature branch. Fixed conflicts and typeerrors

Commit:
Update commandlogspec tests

Commit:
Add commandlogs to import/export process

Commit:
Update import/export with transaction log data

mankyKitty commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,208 @@
+  logFilePath <- failWith ImportWalletError_InvalidCommandLogDestination $
+    _transactionLogger_destination txLogger
+
+  _ <- liftIO $ T.writeFile logFilePath rawCmdLogData
```
Needs to handle exceptions

mankyKitty commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,208 @@
+  cmdLogFile <- failWith ExportWalletError_CommandLogExport $
+    _transactionLogger_destination txLogger
+
+  cmdLogs <- liftIO $ TL.readFile cmdLogFile
```
Needs to handle exceptions

mankyKitty committed:
Delete .Impl.hs.kak.aeJjAf

Junk file from editor.

mankyKitty commented on frontend/src/Frontend/UI/Dialogs/TxLogs.hs:
```diff
@@ -0,0 +1,126 @@
+          let sender = Api._commandLog_sender cmdLog
+          td "tx-log-row__sender" $ case (textToKey sender) :: Maybe PublicKey of
+            Nothing -> text sender
+            Just _ -> text $ Text.take 4 sender <> "..." <> Text.takeEnd 4 sender
```
Change to be first 8 characters of the key, as per discussions and keeps it consistent with export process.

mankyKitty commented on frontend/src/Pact/Server/ApiClient.hs:
```diff
@@ -194,7 +196,49 @@ runTransactionLoggerT :: TransactionLoggerT m a -> TransactionLogger -> m a
+createCommandLogExportFileV1 fp = runExceptT $ do
+  let logFilePath = fp
+      exportFilePath = printf "%s_v%i" fp commandLogCurrentVersion
+  xs <- ExceptT $ traverse Aeson.eitherDecodeStrict . BS8.lines <$> BS.readFile logFilePath
```
Handle exceptions

Commit:
Some fixes for the PR based on review

Commit:
Add more exception handling for log file reading

Commit:
Merge branch 'develop' into sc@merge-tx-log-importexport

alexfmpe commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,234 @@
+
+  -- TODO: Test that the password is good for the XPrv.Crypto
+  let
+    testMsg = "test" :: ByteString
```
Should this really stay here?

Commit:
Remove TODO as it was fixed

alexfmpe commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,234 @@
+  pure $
+    ( fold
+      [ (T.unpack . (T.take 8) . keyToText . _keyPair_publicKey . _key_pair $ keyPair)
+      , "."
```
`T.intercalate "."`

alexfmpe commented on desktop/src/Desktop/Frontend.hs:
```diff
@@ -201,6 +220,17 @@ bipWallet mkAppCfg = do
+                  eFileDone <- _fileFFI_deliverFile frontendFileFFI eGoodExport
+                  pure $
+                    (Left <$> eErrExport)
+                    <> (first ExportWalletError_FileNotWritable <$> eFileDone)
```
This technically requires the `Semigroup Either` instance, which is (I think) irrelevant.
These events seem mutually exclusive, can we make this `leftmost` or `<!>` ?

alexfmpe commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,234 @@
+  -> Value
+  -> ExceptT ImportWalletError m StorageVersion
+extractImportVersionField key ver =
+  hoistParser key ver (withObject "ChainWeaverImport" (\o -> o .:? key .!= 0 ))
```
Odd casing - shouldn't it be `ChainweaverImport` ?

alexfmpe replied:
Don't we always put the version there? 
Seems like a missing version would be unexpected, in which case better fail fast.

alexfmpe commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,234 @@
+  -> Value
+  -> ExceptT ImportWalletError m a
+extractImportDataField key ver =
+  hoistParser key ver (withObject "ChainWeaverImport" (\o -> o .: key))
```
Better factor our this string like the others.

alexfmpe commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,234 @@
+  -> (Value -> Parser a)
+  -> Value
+  -> ExceptT ImportWalletError m a
+hoistParser errLabel ver p  =
```
Is `errLabel` ever not `key` ?

alexfmpe commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,234 @@
+     , Reflex t
+     )
+  => TransactionLogger
+  -> Password -- Password
```
Comment not needed anymore

alexfmpe commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,234 @@
+
+  _ <- ExceptT $ runBIPCryptoT (constant (rootKey, unPassword pw)) $ do
+    let vStore = FrontendStore.versionedStorage
+    feLatestEither <- first (expandDecodeVersionJsonError storeFrontendDataKey feVer)
+      <$> (_versionedStorage_decodeVersionedJson vStore feVer feData)
```
If you'd like, it seems possible to get rid of `expandDecodeVersionJsonError` by flipping this around:
`ffor (_versionedStorage_decodeVersionedJson vStore feVer feData) $ first $ \section expectedVer -> \case`

alexfmpe commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,234 @@
+      <$> (_versionedStorage_decodeVersionedJson vStore feVer feData)
+
+    traverse
+      (\dmap -> liftIO (TL.putStrLn $ encodeToLazyText dmap) >> _versionedStorage_restoreBackup vStore dmap)
```
Why print?

mankyKitty replied:
Because `Show` is for debugging, if we use a specific encoding then we have consistent output rather than raw Haskell data structures that are half pretty printing and half auto-derived.

alexfmpe commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,234 @@
+        Just cmdLogVer -> do
+          rawCmdLogData <- extractImportDataField @Text commandLogDataKey natVer jVal
+          _ <- hoistEither $ first (ImportWalletError_DecodeError commandLogDataKey cmdLogVer . T.pack)
+            $ traverse (eitherDecode @CommandLog . TL.encodeUtf8 . TL.fromStrict) $ T.lines rawCmdLogData
```
`traverse_`

alexfmpe commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,234 @@
+            pure $ Left ImportWalletError_CommandLogWriteError
+
+-- This is running on the assumption that the storage has been upgraded already and isn't on
+-- an old version.
```
Can we enforce this?

Commit:
Update password test message and refactor.

Refactor password verification to a single function.

alexfmpe commented on desktop/src/Desktop/ImportExport.hs:
```diff
@@ -0,0 +1,234 @@
+
+  keyPair <- failWith ExportWalletError_NoKeys $ do
+    (Identity keyMap) <- DMap.lookup StoreFrontend_Wallet_Keys feData
+    IntMap.lookup 0 keyMap
```
Is this different than testing `null keyMap` ?

mankyKitty replied:
We want `keyPair` don't we? It's not just a test.

alexfmpe replied:
Oh nevermind

alexfmpe commented on desktop/src/Desktop/Setup.hs:
```diff
@@ -229,34 +235,42 @@ splashLogo = do
+        create <- confirmButton (def & disabledCfg ) "Create a new wallet"
+        restoreBipPhrase <- uiButtonDyn (btnCfgSecondary & disabledCfg & restoreCfg) $ text "Restore from recovery phrase"
+        restoreImport <- uiButtonDyn (btnCfgSecondary & disabledCfg & restoreCfg) $ text "Restore from wallet export"
+        pure (agreed, create, restoreBipPhrase, restoreImport)
```
Might as well do the `leftmost` here since that's the only reason for plumbing around these `Event`s

alexfmpe commented on desktop/src/Desktop/Setup.hs:
```diff
@@ -323,10 +337,92 @@ recoverWallet eBack = Workflow $ do
+restoreFromImport fileFFI backWF eBack = nagScreen
+  where
+    nagBack = uiButtonDyn
+      -- TODO: Don't reuse this class or at least rename it
```
Ok

alexfmpe committed:
Merge branch 'develop' into sc@merge-tx-log-importexport

Commit:
Fix back behaviour on import when wallet is empty

alexfmpe commented on frontend/src/Frontend.hs:
```diff
@@ -58,39 +61,58 @@ frontend = Frontend
+          liftIO $ cb $ ((name,) <$> join mText)
+        pure ()
+      let open = do
+            -- This doesn't fix the bug where if you open the same file twice we don't get an event
```
Bugger - ticket?

mankyKitty replied:
Done

alexfmpe commented on frontend/src/Frontend/AppCfg.hs:
```diff
@@ -17,16 +17,45 @@ data ChangePassword key t m = ChangePassword
+  -- ^ Delivers a file to an appropriate place to the user
   }
 
+-- TODO: Tidy this up
```
In what sense?

mankyKitty replied:
You'd have to ask @benkolera , removing the TODO as it looks pretty damn tidy to me.

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/DeployConfirmation.hs:
```diff
@@ -246,6 +248,22 @@ listenToRequestKey clientEnvs onRequestKey = do
               setMessage $ Just $ Left $ NetworkError_CommandFailure err
             -- This case shouldn't happen
             _ -> listen Status_Failed
+-- <<<<<<< HEAD
```
Uh oh

mankyKitty replied:
Merge snafu that isn't as bad as it looks, backtracked it a bit and the original change came from https://github.com/kadena-io/chainweaver/commit/368fe8f10b84ef79c33872fc44e00e4e14f1fa3b but in all the rebasing/merging it appeared as change again and just wasn't cleaned up. Boo, my bad.

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/ExportWallet.hs:
```diff
@@ -0,0 +1,94 @@
+            <> "class" =: "input_width_full"
+            )
+        let errWidget = elClass "p" "error_inline" . text
+        _ <- widgetHold blank $ ffor eErr $ \case
```
Maybe 
```haskell
widgetHold_ blank $ ffor eErr $ elClass "p" "error_inline" . text . \case
```

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/TxLogs.hs:
```diff
@@ -0,0 +1,120 @@
+  -> model
+  -> Event t ()
+  -> m (mConf, Event t ())
+uiTxLogs fileFFI _model _onExtClose = do
```
Why keep `_model` ?

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/TxLogs.hs:
```diff
@@ -0,0 +1,120 @@
+      onDeliveredFileOk <- _fileFFI_deliverFile fileFFI onContentsReady
+
+      onClick <- uiButtonDyn cfg' $ do
+        dynText $ ffor (cfg' ^. uiButtonCfg_title) fold
```
What's the effect of the `title` attribute? Why not simply have `text "Export Full Transaction Log"` ?

mankyKitty replied:
mea copy-pasta

alexfmpe commented on frontend/src/Frontend/UI/Settings.hs:
```diff
@@ -32,21 +34,33 @@ type HasUiSettingModelCfg model mConf key m t =
     [ ffor (_enabledSettings_changePassword enabledSettings) $ \changePassword -> do
       settingItem "Change Password" (static @"img/lock-light.svg") (uiChangePasswordDialog changePassword)
+    , ffor (_enabledSettings_exportWallet enabledSettings) $ \exportWallet-> do
+      -- TODO: Need to center the svg properly
```
Seems centered enough to me?

alexfmpe commented on frontend/src/Frontend/VersionedStore.hs:
```diff
@@ -0,0 +1,149 @@
+  where
+    -- Should we do this, or should we just push on? The backup functionality really
+    -- depends on the storage being properly latest, but I think that our failthrough
+    -- to nothing behaviour in storage will just zero everything out to latest version...
```
Let's chat about this

alexfmpe commented on frontend/src/Frontend/VersionedStore.hs:
```diff
@@ -0,0 +1,149 @@
+    restoreBackup :: DMap (Latest.StoreFrontend key) Identity -> m ()
+    restoreBackup dm = restoreLocalStorageDump prefix dm 1
+
+    -- Takes a json blob and upgrades it the the latest DMap structure
```
`to the`

alexfmpe commented on mac/src/Mac.hs:
```diff
@@ -37,11 +40,19 @@ ffi = AppFFI
   , _appFFI_moveToForeground = moveToForeground
   , _appFFI_resizeWindow = uncurry resizeWindow
-  , _appFFI_global_openFileDialog = global_openFileDialog
+  -- TODO : Take file type and give it to the dialog
```
Isn't this done?

alexfmpe commented:
Mac build fails for me when running tests of `system-locale`

Commit:
Updates from PR feedback

Commit:
Force an upgrade on wallet export.

Should be a no-op in all cases as loading the wallet will upgrade your data. However if we
add other ways to create the export then this may be forgotten.

Commit:
Fix mac build and correct typo in export wallet message

Commit:
Had the override in the wrong spot, added comment

Commit:
Don't fail on empty command logs on import
